File: app/src/main/java/com/lifedawn/bestweather/weathers/view/DateView.java
package com.lifedawn.bestweather.weathers.view;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.text.StaticLayout;
import android.text.TextPaint;
import android.util.TypedValue;
import android.view.View;

import com.lifedawn.bestweather.R;
import com.lifedawn.bestweather.theme.AppTheme;
import com.lifedawn.bestweather.weathers.FragmentType;

import java.time.ZonedDateTime;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.List;

public class DateView extends View {
	private final FragmentType fragmentType;

	private final TextPaint dateTextPaint;
	private final int viewWidth;
	private final int columnWidth;
	private final DateTimeFormatter dateTimeFormatter;
	private List<DateValue> dateValueList;
	private int currentX;
	private int firstColX;
	private int textHeight;

	private int textSize;
	private Rect rect = new Rect();
	private int padding;
	private int viewHeight;

	public DateView(Context context, FragmentType fragmentType, int viewWidth, int columnWidth) {
		super(context);
		this.fragmentType = fragmentType;
		this.viewWidth = viewWidth;
		this.columnWidth = columnWidth;
		dateTimeFormatter = DateTimeFormatter.ofPattern("M.d\nE");
		padding = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 6f, getResources().getDisplayMetrics());

		dateTextPaint = new TextPaint(Paint.ANTI_ALIAS_FLAG);
		dateTextPaint.setTextAlign(Paint.Align.CENTER);
		dateTextPaint.setTextSize(context.getResources().getDimension(R.dimen.dateValueTextSizeInSCD));
		dateTextPaint.setColor(AppTheme.getTextColor(context, fragmentType));

		setWillNotDraw(false);
	}

	public void setTextSize(int textSizeSp) {
		this.textSize = (int) TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_SP, textSizeSp, getResources().getDisplayMetrics());
		dateTextPaint.setTextSize(textSize);
	}

	public void init(List<ZonedDateTime> dateTimeList) {
		ZonedDateTime date = ZonedDateTime.of(dateTimeList.get(0).toLocalDateTime(), dateTimeList.get(0).getZone());
		ZonedDateTime lastDate = ZonedDateTime.of(date.toLocalDateTime(), date.getZone());

		List<DateView.DateValue> dateValueList = new ArrayList<>();
		int beginX = 0;

		for (int col = 0; col < dateTimeList.size(); col++) {
			date = ZonedDateTime.of(dateTimeList.get(col).toLocalDateTime(), lastDate.getZone());

			if (date.getHour() == 0 || col == 0) {
				if (dateValueList.size() > 0) {
					dateValueList.get(dateValueList.size() - 1).endX = columnWidth * (col - 1) + columnWidth / 2;
				}
				beginX = columnWidth * col + columnWidth / 2;
				dateValueList.add(new DateView.DateValue(beginX, date));
			}

			if (lastDate.getDayOfYear() != date.getDayOfYear()) {
				lastDate = date;
			}
		}
		dateValueList.get(dateValueList.size() - 1).endX = columnWidth * (dateTimeList.size() - 1) + columnWidth / 2;

		this.dateValueList = dateValueList;
		this.firstColX = dateValueList.get(0).beginX;
	}


	@Override
	protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
		ZonedDateTime now = ZonedDateTime.now();
		String val = now.format(dateTimeFormatter);
		dateTextPaint.getTextBounds(val, 0, val.length(), rect);
		StaticLayout.Builder builder = StaticLayout.Builder.obtain(val, 0, val.length(), dateTextPaint, columnWidth);
		StaticLayout sl = builder.build();
		viewHeight = rect.height() * sl.getLineCount() + padding * 2;

		setMeasuredDimension(viewWidth, viewHeight);
	}

	@Override
	protected void onLayout(boolean changed, int left, int top, int right, int bottom) {
		super.onLayout(changed, left, top, right, bottom);
	}

	@Override
	protected void onDraw(Canvas canvas) {
		super.onDraw(canvas);
		final int y = getHeight() / 2 + textHeight / 2;

		for (DateValue dateValue : dateValueList) {
			if (currentX >= dateValue.beginX - firstColX && currentX < dateValue.endX - firstColX) {
				dateValue.lastX = currentX + firstColX;
			} else if (currentX < dateValue.beginX) {
				dateValue.lastX = dateValue.beginX;
			}
			//canvas.drawText(dateValue.date.format(dateTimeFormatter), dateValue.lastX, y, dateTextPaint);
			drawText(canvas, dateValue.date.format(dateTimeFormatter), dateValue.lastX);
		}
	}

	public void reDraw(int newX) {
		this.currentX = newX;
		invalidate();
	}

	private void drawText(Canvas canvas, String textOnCanvas, float x) {
		dateTextPaint.getTextBounds(textOnCanvas, 0, textOnCanvas.length(), rect);
		StaticLayout.Builder builder = StaticLayout.Builder.obtain(textOnCanvas, 0, textOnCanvas.length(), dateTextPaint, columnWidth);
		StaticLayout sl = builder.build();

		canvas.save();

		float textHeight = rect.height();
		float textYCoordinate = viewHeight / 2f + rect.exactCenterY() -
				((sl.getLineCount() * textHeight) / 2);

		final float columnCenterX = columnWidth / 2f;
		float textXCoordinate = columnCenterX + rect.left;

		canvas.translate(x, textYCoordinate);

		sl.draw(canvas);
		canvas.restore();
	}


	public static class DateValue {
		public final int beginX;
		public final ZonedDateTime date;
		public int endX;
		public int lastX;

		public DateValue(int beginX, ZonedDateTime date) {
			this.beginX = beginX;
			this.date = date;
			this.lastX = beginX;
		}
	}
}

File: app/src/main/java/com/lifedawn/bestweather/ui/weathers/customview/DateView.kt
package com.lifedawn.bestweather.ui.weathers.customview

import android.content.Context
import android.graphics.Canvas
import android.graphics.Paint
import android.text.StaticLayout
import android.text.TextPaint
import android.util.TypedValue
import android.view.View
import com.lifedawn.bestweather.R
import com.lifedawn.bestweather.ui.theme.AppTheme.getTextColor
import com.lifedawn.bestweather.ui.weathers.enums.WeatherDataType
import java.time.ZonedDateTime
import java.time.format.DateTimeFormatter

class DateView(context: Context, private val weatherDataType: WeatherDataType, private val viewWidth: Int, private val columnWidth: Int) :
    View(context), ICleaner {
    private val dateTextPaint: TextPaint
    private val dateTimeFormatter = DateTimeFormatter.ofPattern("M.d\nE")
    private var dateValueList: MutableList<DateValue>? = null
    private var currentX = 0
    private var firstColX = 0
    private val padding: Int
    private var viewHeight = 0

    init {
        padding = TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, 2f, resources.displayMetrics).toInt()
        dateTextPaint = TextPaint(Paint.ANTI_ALIAS_FLAG)
        dateTextPaint.textAlign = Paint.Align.CENTER
        dateTextPaint.textSize = context.resources.getDimension(R.dimen.dateValueTextSizeInSCD)
        dateTextPaint.color = getTextColor(weatherDataType)
        setWillNotDraw(false)
    }

    fun setTextSize(textSizeSp: Int) {
        dateTextPaint.textSize = TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_SP, textSizeSp.toFloat(),
            resources.displayMetrics
        ).toInt().toFloat()
    }

    fun setTextColor(textColor: Int) {
        dateTextPaint.color = textColor
    }

    fun init(dateTimeList: List<ZonedDateTime>) {
        var date = ZonedDateTime.of(dateTimeList[0].toLocalDateTime(), dateTimeList[0].zone)
        var lastDate = ZonedDateTime.of(date.toLocalDateTime(), date.zone)
        lastDate = lastDate.minusDays(5)
        val dateValueList: MutableList<DateValue> = ArrayList()
        var beginX = 0
        for (col in dateTimeList.indices) {
            date = ZonedDateTime.of(dateTimeList[col].toLocalDateTime(), lastDate.zone)
            if (date.dayOfYear != lastDate.dayOfYear || col == 0) {
                if (dateValueList.size > 0) {
                    dateValueList[dateValueList.size - 1].endX = columnWidth * (col - 1) + columnWidth / 2
                }
                beginX = columnWidth * col + columnWidth / 2
                dateValueList.add(DateValue(beginX, date))
                lastDate = date
            }
        }
        dateValueList[dateValueList.size - 1].endX = columnWidth * (dateTimeList.size - 1) + columnWidth / 2
        this.dateValueList = dateValueList
        firstColX = dateValueList[0].beginX
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        var staticLayout: StaticLayout? = null
        viewHeight = Int.MIN_VALUE
        var str: String? = null
        for (`val` in dateValueList!!) {
            str = `val`.date.format(dateTimeFormatter)
            val builder = StaticLayout.Builder.obtain(
                str, 0, str.length, dateTextPaint,
                columnWidth
            )
            staticLayout = builder.build()
            viewHeight = Math.max(staticLayout.height, viewHeight)
        }
        viewHeight += padding * 2
        setMeasuredDimension(viewWidth, viewHeight)
    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)
        for (dateValue in dateValueList!!) {
            if (currentX >= dateValue.beginX - firstColX && currentX < dateValue.endX - firstColX) {
                dateValue.lastX = currentX + firstColX
            } else if (currentX < dateValue.beginX) {
                dateValue.lastX = dateValue.beginX
            }
            drawText(canvas, dateValue.date.format(dateTimeFormatter), dateValue.lastX.toFloat())
        }
    }

    fun reDraw(newX: Int) {
        currentX = newX
        invalidate()
    }

    private fun drawText(canvas: Canvas, textOnCanvas: String, x: Float) {
        val builder = StaticLayout.Builder.obtain(textOnCanvas, 0, textOnCanvas.length, dateTextPaint, columnWidth)
        val staticLayout = builder.build()
        val y = viewHeight / 2f - staticLayout.height / 2f
        canvas.save()
        canvas.translate(x, y)
        staticLayout.draw(canvas)
        canvas.restore()
    }

    override fun clear() {
        if (dateValueList != null) dateValueList!!.clear()
        dateValueList = null
    }

    class DateValue(val beginX: Int, val date: ZonedDateTime) {
        var endX = 0
        var lastX: Int

        init {
            lastX = beginX
        }
    }
}